core/src/main/java/com/linecorp/armeria/internal/ObjectsToHttpResponseConvertingSubscriber.java

```diff
@@ -0,0 +1,121 @@
+import com.linecorp.armeria.server.annotation.ExceptionHandlerFunction;
+import com.linecorp.armeria.server.annotation.ResponseConverterFunction;
+
+/**
+ * A {@link Subscriber} which collects all objects produced by a {@link Publisher} as a list.
+ * The collected objects would be converted to an {@link HttpResponse} using the specified
+ * {@link ResponseConverterFunction} when {@link #onComplete()} is called, then the {@link HttpResponse}
+ * would complete the {@link CompletableFuture} which is given when creating this instance.
+ */
+public class ObjectsToHttpResponseConvertingSubscriber implements Subscriber<Object> {
+
+    private final ServiceRequestContext ctx;
+    private final HttpRequest req;
+    private final CompletableFuture<HttpResponse> future;
+    private final ResponseConverterFunction configuredResponseConverter;
+    private final ExceptionHandlerFunction configuredExceptionHandler;
+
+    @Nullable
+    private List<Object> objects;
+
+    /**
+     * Creates a new instance.
+     *
+     * @param ctx the {@link ServiceRequestContext} of the {@code req}
+     * @param req the {@link HttpRequest} received by the server
+     * @param future the {@link CompletableFuture} which waits for an {@link HttpResponse}
+     * @param configuredResponseConverter the {@link ResponseConverterFunction} which is used to convert the
+     *                                    collected objects into an {@link HttpResponse} when the
+     *                                    {@link #onComplete()} is invoked
+     * @param configuredExceptionHandler the {@link ExceptionHandlerFunction} which is used to convert the
+     *                                   {@link Throwable} into an {@link HttpResponse} when the
+     *                                   {@link #onError(Throwable)} is invoked
+     */
+    public ObjectsToHttpResponseConvertingSubscriber(ServiceRequestContext ctx, HttpRequest req,
+                                                     CompletableFuture<HttpResponse> future,
+                                                     ResponseConverterFunction configuredResponseConverter,
+                                                     ExceptionHandlerFunction configuredExceptionHandler) {
+        this.ctx = requireNonNull(ctx, "ctx");
+        this.req = requireNonNull(req, "req");
+        this.future = requireNonNull(future, "future");
+        this.configuredResponseConverter = requireNonNull(configuredResponseConverter,
+                                                          "configuredResponseConverter");
+        this.configuredExceptionHandler = requireNonNull(configuredExceptionHandler,
+                                                         "configuredExceptionHandler");
+    }
+
+    @Override
+    public void onSubscribe(Subscription s) {
+        s.request(Integer.MAX_VALUE);
+    }
+
+    @Override
+    public void onNext(Object o) {
+        if (objects == null) {
+            objects = new ArrayList<>();
```

Comment:
How about:

- using `ImmutableList.Builder`?
- having a dedicated field for the first element so that we don't need to create a list when the stream contains only one element? e.g.
  ```java
  private Object firstElement;
  private ImmutableList.Builder<Object> listBuilder;

  public void onNext(Object o) {
      if (firstElement == null) {
          firstElement = o;
      } else {
          if (listBuilder != null) {
              listBuilder = ImmutabletList.builder();
              listBuilder.add(firstElement);
          }
          listBuilder.add(o);
      }
  }

  public void onComplete() {
      final Object obj;
      if (listBuilder != null) {
          obj = listBuilder.build();
          listBuilder = null;
      } else {
          obj = firstElement;
      }
      firstElement = null;
      ...
  }
  ```